utils/checks.stanza

```diff
@@ -113,8 +133,207 @@ public pcb-check rated-temperature (component:JITXObject):
+          open-collector-checks(out-p, o, all-connected-pins)
+ 
+; Checks any IO pins marked CMOSOutput or TLLOutput
+defn cmos-and-ttl-checks (out-p:JITXObject,
+                          connected-io-pins) : 
+  for in-p in filter({_ != out-p}, connected-io-pins) do :
+    ; run the checks
+    check-single-push-pull-levels(out-p, in-p)
+
+; Checks that run on a pin marked OpenCollector
+; - io-pin: the pin we are checking
+; - driver: the driver (value of driver(property(io-pin.digital-io)))
+; - all-connected-pins: the pins this pin is connected to
+;
+defn open-collector-checks (io-pin:JITXObject, 
+                            driver:OpenCollector,
+                            all-connected-pins:Tuple<JITXObject>) : 
+  ; First, we search for the pull-up-resistors that might be connected to this pin
+  val pull-up-resistors = to-tuple $
+    for p in all-connected-pins seq? : 
+      val instance = containing-instance(p) as Instance
+      val connected-to-rail? = 
+        for p in pins(instance) any? : 
+          has-property?(p.rail-voltage)
+      if has-property?(instance.resistance) and connected-to-rail? : 
+        One(instance) 
+      else : 
+        None()
+
+  ; Check that there exists at least one pull-up-resistor
+  #CHECK(
+    condition =            not empty?(pull-up-resistors),
+    name =                 "IO pin checks"
+    description =          "Check io pin properties"
+    category =             "pin-check"
+    subcheck-description = "Check that io pins marked OpenCollector have pullup resistors",
+    pass-message =         "Pin %_ is marked OpenCollector and has pull-up resistor(s)" % [ref(io-pin)],
+    fail-message =         "Pin %_ is marked OpenCollector and is missing pull-up resistor(s)" % [ref(io-pin)],
+    locators =             [io-pin])
+
+  ; Next, we check that all the pull-up resistors are connected to the same rail
+  val same-rail-voltage? = all-equal? $ 
+    for resistor in pull-up-resistors seq :
+      for p in pins(resistor) first! : 
+        property?(p.rail-voltage)
+    
+  #CHECK(
+    condition =            same-rail-voltage?,
+    name =                 "IO pin checks"
+    description =          "Check io pin properties"
+    category =             "pin-check"
+    subcheck-description = "Check that io pins marked OpenCollector have pullup resistors connected to the same rail voltage",
+    pass-message =         "Pin %_ is marked OpenCollector and has pull-up resistor(s) with correct rail voltage" % [ref(io-pin)],
+    fail-message =         "Pin %_ is marked OpenCollector and pull-up resistors have different rail voltages" % [ref(io-pin)],
+    locators =             [io-pin])
+
+  ; Here we find the value of the rail voltage
+  val rail-voltage = let : 
+    val r = pull-up-resistors[0]
+    for p in pins(r) first! :   
+      property?(p.rail-voltage)
+
+  ; And the total resistance of parallel connected
+  ; pull-up resistors
+  val total-resistance = 1.0 / sum $
```

Comment:
This is SO cool to calculate the parallel value of resistance using stanza syntax!